pylsp/plugins/jedi_completion.py

```diff
@@ -62,36 +62,34 @@ def pylsp_completions(config, document, position):
         SNIPPET_RESOLVER.cached_modules = modules_to_cache_for
 
     include_params = snippet_support and should_include_params and use_snippets(document, position)
-    include_class_objects = snippet_support and should_include_class_objects and use_snippets(document, position)
-
-    ready_completions = [
-        _format_completion(
-            c,
-            include_params,
-            resolve=resolve_eagerly,
-            resolve_label_or_snippet=(i < max_to_resolve)
-        )
-        for i, c in enumerate(completions)
-    ]
-
-    # TODO split up once other improvements are merged
-    if include_class_objects:
-        for i, c in enumerate(completions):
-            if c.type == 'class':
-                completion_dict = _format_completion(
+    include_class_objects = (
+        snippet_support and should_include_class_objects and use_snippets(document, position)
+    )
+
+    ready_completions = []
+    for i, c in enumerate(completions):
+
+        ready_completions.append(
+            {
+                **_format_completion(
                     c,
-                    False,
+                    include_params,
                     resolve=resolve_eagerly,
-                    resolve_label_or_snippet=(i < max_to_resolve)
-                )
-                completion_dict['kind'] = lsp.CompletionItemKind.TypeParameter
-                completion_dict['label'] += ' object'
-                ready_completions.append(completion_dict)
+                    resolve_label_or_snippet=(i < max_to_resolve),
+                ),
+                'data': {'doc_uri': document.uri},
+            }
+        )
+
+        if c.type == 'class' and include_class_objects:
+            completion_dict = _format_completion(
+                c, False, resolve=resolve_eagerly, resolve_label_or_snippet=(i < max_to_resolve)
+            )
+            completion_dict['kind'] = lsp.CompletionItemKind.TypeParameter
+            completion_dict['label'] += ' object'
+            completion_dict['data'] = {'doc_uri': document.uri}
+            ready_completions.append(completion_dict)
 
-    for completion_dict in ready_completions:
-        completion_dict['data'] = {
-            'doc_uri': document.uri
-        }
```

Comment:
Merging this one into the two loops above is a good idea. I would propose to limit this PR to this change only. We can consider other changes together with benchamrking/profiling results to evidence those are actually beneficial.

Reply:
I wonder if instead of creating the dictionary _n_ times, we could just create it once outside of the loop and assign the same instance to all objects; this might bite us in the future or in other places where we may assume this object to be mutable though.